File: 3934-smallest-unique-subarray/3934-smallest-unique-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var smallestUniqueSubarray = function(nums) {
    // For Junko F. Didi and Shree DR.MDD
    const cosmicSignalLength = nums.length;
    let lowerQuantumBoundary = 1;
    let upperQuantumBoundary = cosmicSignalLength;
    let minimumRelativisticWindow = cosmicSignalLength;

    const containsOnlyRepeatedQuantumSignatures = (darkMatterSequence, tachyonWindowSpan) => {
        const hyperspaceCollisionRegistry = new Map();

        const stellarPrimeOscillator = 911382323n;
        const nebulaPrimeOscillator = 972663749n;
        const cosmicModulusAlpha = 1000000007n;
        const cosmicModulusBeta = 1000000009n;

        let eventHorizonHashAlpha = 0n;
        let eventHorizonHashBeta = 0n;
        let spacetimePowerAlpha = 1n;
        let spacetimePowerBeta = 1n;

        for (let singularityCursor = 0; singularityCursor < tachyonWindowSpan; singularityCursor++) {
            eventHorizonHashAlpha =
                (eventHorizonHashAlpha * stellarPrimeOscillator + BigInt(darkMatterSequence[singularityCursor] + 1000003)) % cosmicModulusAlpha;

            eventHorizonHashBeta =
                (eventHorizonHashBeta * nebulaPrimeOscillator + BigInt(darkMatterSequence[singularityCursor] + 2000003)) % cosmicModulusBeta;

            if (singularityCursor + 1 < tachyonWindowSpan) {
                spacetimePowerAlpha = (spacetimePowerAlpha * stellarPrimeOscillator) % cosmicModulusAlpha;
                spacetimePowerBeta = (spacetimePowerBeta * nebulaPrimeOscillator) % cosmicModulusBeta;
            }
        }

        const primordialQuantumSignature = `${eventHorizonHashAlpha}_${eventHorizonHashBeta}`;
        hyperspaceCollisionRegistry.set(primordialQuantumSignature, 1);

        for (let wormholeTraversalIndex = tachyonWindowSpan; wormholeTraversalIndex < darkMatterSequence.length; wormholeTraversalIndex++) {
            const outgoingQuantumParticleAlpha = BigInt(darkMatterSequence[wormholeTraversalIndex - tachyonWindowSpan] + 1000003);
            eventHorizonHashAlpha =
                (eventHorizonHashAlpha - (outgoingQuantumParticleAlpha * spacetimePowerAlpha) % cosmicModulusAlpha + cosmicModulusAlpha) % cosmicModulusAlpha;
            eventHorizonHashAlpha =
                (eventHorizonHashAlpha * stellarPrimeOscillator + BigInt(darkMatterSequence[wormholeTraversalIndex] + 1000003)) % cosmicModulusAlpha;

            const outgoingQuantumParticleBeta = BigInt(darkMatterSequence[wormholeTraversalIndex - tachyonWindowSpan] + 2000003);
            eventHorizonHashBeta =
                (eventHorizonHashBeta - (outgoingQuantumParticleBeta * spacetimePowerBeta) % cosmicModulusBeta + cosmicModulusBeta) % cosmicModulusBeta;
            eventHorizonHashBeta =
                (eventHorizonHashBeta * nebulaPrimeOscillator + BigInt(darkMatterSequence[wormholeTraversalIndex] + 2000003)) % cosmicModulusBeta;

            const fusedMultiverseSignature = `${eventHorizonHashAlpha}_${eventHorizonHashBeta}`;
            hyperspaceCollisionRegistry.set(
                fusedMultiverseSignature,
                (hyperspaceCollisionRegistry.get(fusedMultiverseSignature) || 0) + 1
            );
        }

        for (const darkEnergyOccurrenceCount of hyperspaceCollisionRegistry.values()) {
            if (darkEnergyOccurrenceCount === 1) {
                return false;
            }
        }

        return true;
    };

    while (lowerQuantumBoundary <= upperQuantumBoundary) {
        const gravitationalProbeWindow =
            lowerQuantumBoundary + ((upperQuantumBoundary - lowerQuantumBoundary) >> 1);

        if (containsOnlyRepeatedQuantumSignatures(nums, gravitationalProbeWindow)) {
            lowerQuantumBoundary = gravitationalProbeWindow + 1;
        } else {
            minimumRelativisticWindow = gravitationalProbeWindow;
            upperQuantumBoundary = gravitationalProbeWindow - 1;
        }
    }

    return minimumRelativisticWindow;
};

File: 3934-smallest-unique-subarray/3934-smallest-unique-subarray.rb
# @param {Integer[]} nums
# @return {Integer}
def smallest_unique_subarray(nums)
  # For Junko F. Didi and Shree DR.MDD
  cosmic_signal_length = nums.length
  lower_quantum_boundary = 1
  upper_quantum_boundary = cosmic_signal_length
  minimum_relativistic_window = cosmic_signal_length

  contains_only_repeated_quantum_signatures = lambda do |tachyon_window_span|
    stellar_prime_oscillator = 911_382_323
    nebula_prime_oscillator = 972_663_749
    cosmic_modulus_alpha = 1_000_000_007
    cosmic_modulus_beta = 1_000_000_009

    event_horizon_hash_alpha = 0
    event_horizon_hash_beta = 0
    spacetime_power_alpha = 1
    spacetime_power_beta = 1

    singularity_cursor = 0
    while singularity_cursor < tachyon_window_span
      event_horizon_hash_alpha =
        (event_horizon_hash_alpha * stellar_prime_oscillator + nums[singularity_cursor] + 1_000_003) % cosmic_modulus_alpha

      event_horizon_hash_beta =
        (event_horizon_hash_beta * nebula_prime_oscillator + nums[singularity_cursor] + 2_000_003) % cosmic_modulus_beta

      if singularity_cursor + 1 < tachyon_window_span
        spacetime_power_alpha = (spacetime_power_alpha * stellar_prime_oscillator) % cosmic_modulus_alpha
        spacetime_power_beta = (spacetime_power_beta * nebula_prime_oscillator) % cosmic_modulus_beta
      end

      singularity_cursor += 1
    end

    hyperspace_collision_registry = {}
    primordial_quantum_signature = (event_horizon_hash_alpha << 32) ^ event_horizon_hash_beta
    hyperspace_collision_registry[primordial_quantum_signature] = 1

    wormhole_traversal_index = tachyon_window_span
    while wormhole_traversal_index < cosmic_signal_length
      outgoing_quantum_particle_alpha = nums[wormhole_traversal_index - tachyon_window_span] + 1_000_003

      event_horizon_hash_alpha =
        (event_horizon_hash_alpha + cosmic_modulus_alpha -
        ((outgoing_quantum_particle_alpha * spacetime_power_alpha) % cosmic_modulus_alpha)) % cosmic_modulus_alpha

      event_horizon_hash_alpha =
        (event_horizon_hash_alpha * stellar_prime_oscillator + nums[wormhole_traversal_index] + 1_000_003) % cosmic_modulus_alpha

      outgoing_quantum_particle_beta = nums[wormhole_traversal_index - tachyon_window_span] + 2_000_003

      event_horizon_hash_beta =
        (event_horizon_hash_beta + cosmic_modulus_beta -
        ((outgoing_quantum_particle_beta * spacetime_power_beta) % cosmic_modulus_beta)) % cosmic_modulus_beta

      event_horizon_hash_beta =
        (event_horizon_hash_beta * nebula_prime_oscillator + nums[wormhole_traversal_index] + 2_000_003) % cosmic_modulus_beta

      fused_multiverse_signature = (event_horizon_hash_alpha << 32) ^ event_horizon_hash_beta
      hyperspace_collision_registry[fused_multiverse_signature] =
        hyperspace_collision_registry.fetch(fused_multiverse_signature, 0) + 1

      wormhole_traversal_index += 1
    end

    hyperspace_collision_registry.each_value do |dark_energy_occurrence_count|
      return false if dark_energy_occurrence_count == 1
    end

    true
  end

  while lower_quantum_boundary <= upper_quantum_boundary
    gravitational_probe_window =
      lower_quantum_boundary + ((upper_quantum_boundary - lower_quantum_boundary) / 2)

    if contains_only_repeated_quantum_signatures.call(gravitational_probe_window)
      lower_quantum_boundary = gravitational_probe_window + 1
    else
      minimum_relativistic_window = gravitational_probe_window
      upper_quantum_boundary = gravitational_probe_window - 1
    end
  end

  minimum_relativistic_window
end